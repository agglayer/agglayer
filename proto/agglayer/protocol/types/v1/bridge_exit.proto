syntax = "proto3";

package agglayer.protocol.types.v1;

import "agglayer/protocol/types/v1/bytes.proto";

// Represents a token bridge exit from the network.
message BridgeExit {
  // The type of the leaf.
  LeafType leaf_type = 1;

  // Unique ID for the token being transferred.
  TokenInfo token_info = 2;

  // Network which the token is transferred to
  uint32 dest_network = 3;

  // Address which will own the received token
  FixedBytes20 dest_address = 4;

  // Token amount sent, stored big-endian
  FixedBytes32 amount = 5;

  // Metadata for the bridge exit
  FixedBytes32 metadata = 6;
}

// The type of the leaf.
enum LeafType {
  // Unspecified leaf type.
  LEAF_TYPE_UNSPECIFIED = 0;
  // Transfer leaf type.
  LEAF_TYPE_TRANSFER = 1;
  // Message leaf type.
  LEAF_TYPE_MESSAGE = 2;
}

// Encapsulates the information to uniquely identify a token on the origin
// network.
message TokenInfo {
  // Network which the token originates from
  uint32 origin_network = 1;
  // The address of the token on the origin network
  FixedBytes20 origin_token_address = 2;
}
