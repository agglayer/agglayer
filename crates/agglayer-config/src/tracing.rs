use std::{fmt::Display, path::PathBuf};

use serde::{Deserialize, Deserializer, Serialize};
use tracing_subscriber::{fmt::writer::BoxMakeWriter, EnvFilter};

/// The tracing configuration.
#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct Tracing {
    /// The `RUST_LOG` environment variable will take precedence over the
    /// configuration tracing level.
    #[serde(default)]
    pub level: TracingLevel,
    #[serde(default)]
    pub outputs: Vec<TracingOutput>,
    #[serde(default)]
    pub format: TracingFormat,
    /// Socket of the open telemetry agent endpoint.
    /// If not provided open telemetry will not be used.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub otlp_agent: Option<String>,
    /// Otlp service name.
    /// If not provided open telemetry will not be used.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub otlp_service_name: Option<String>,
}

/// The log format.
#[derive(Serialize, Deserialize, Debug, Default, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum TracingFormat {
    #[default]
    Pretty,
    Json,
}

/// The log level.
#[derive(Serialize, Deserialize, Debug, Default, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum TracingLevel {
    Trace,
    Debug,
    #[default]
    Info,
    Warn,
    Error,
    Fatal,
}

impl Display for TracingLevel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let level = match self {
            TracingLevel::Trace => "trace",
            TracingLevel::Debug => "debug",
            TracingLevel::Info => "info",
            TracingLevel::Warn => "warn",
            TracingLevel::Error => "error",
            TracingLevel::Fatal => "fatal",
        };

        write!(f, "{level}")
    }
}

impl From<TracingLevel> for EnvFilter {
    fn from(value: TracingLevel) -> Self {
        EnvFilter::new(format!("warn,agglayer={value},pessimistic_proof={value}"))
    }
}

/// The log output.
///
/// This can be either `stdout`, `stderr`, or a file path.
///
/// The [`Deserialize`] implementation allows for the configuration file to
/// specify the output location as a string, which is then parsed into the
/// appropriate enum variant. If the string is not recognized to be either
/// `stdout` or `stderr`, it is assumed to be a file path.
#[derive(Serialize, Debug, Clone, Default, PartialEq, Eq)]
pub enum TracingOutput {
    #[default]
    Stdout,
    Stderr,
    File(PathBuf),
    Otlp,
}

impl<'de> Deserialize<'de> for TracingOutput {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        // If the string is not recognized to be either `stdout` or `stderr`,
        // it is assumed to be a file path.
        match s.as_str() {
            "stdout" => Ok(TracingOutput::Stdout),
            "stderr" => Ok(TracingOutput::Stderr),
            "otlp" => Ok(TracingOutput::Otlp),
            _ => Ok(TracingOutput::File(PathBuf::from(s))),
        }
    }
}

impl TracingOutput {
    /// Get a [`BoxMakeWriter`] for the tracing output.
    ///
    /// This can be used to plug the tracing output into the tracing subscriber.
    pub fn as_make_writer(&self) -> BoxMakeWriter {
        match self {
            TracingOutput::Stdout => BoxMakeWriter::new(std::io::stdout),
            TracingOutput::Stderr => BoxMakeWriter::new(std::io::stderr),
            TracingOutput::Otlp => BoxMakeWriter::new(std::io::stdout),
            TracingOutput::File(path) => {
                let appender = tracing_appender::rolling::never(".", path);
                BoxMakeWriter::new(appender)
            }
        }
    }
}
