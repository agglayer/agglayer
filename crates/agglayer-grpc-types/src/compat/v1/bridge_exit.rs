use agglayer_types::{BridgeExit, LeafType, NetworkId};

use super::Error;
use crate::protocol::types::v1;

impl TryFrom<v1::BridgeExit> for BridgeExit {
    type Error = Error;

    fn try_from(value: v1::BridgeExit) -> Result<Self, Self::Error> {
        Ok(BridgeExit {
            leaf_type: match value.leaf_type() {
                v1::LeafType::Transfer => LeafType::Transfer,
                v1::LeafType::Message => LeafType::Message,
                _ => {
                    let t = value.leaf_type;
                    return Err(Error::invalid_data(format!("invalid leaf type: {t}")));
                }
            },
            token_info: required_field!(value, token_info),
            dest_network: NetworkId::new(value.dest_network),
            dest_address: required_field!(value, dest_address),
            amount: required_field!(value, amount),
            metadata: value
                .metadata
                .map(TryInto::try_into)
                .transpose()
                .map_err(|e: Error| e.inside_field("metadata"))?,
        })
    }
}

impl From<BridgeExit> for v1::BridgeExit {
    fn from(value: BridgeExit) -> Self {
        v1::BridgeExit {
            leaf_type: match value.leaf_type {
                LeafType::Transfer => v1::LeafType::Transfer,
                LeafType::Message => v1::LeafType::Message,
            }
            .into(),
            token_info: Some(value.token_info.into()),
            dest_network: value.dest_network.into(),
            dest_address: Some(value.dest_address.into()),
            amount: Some(value.amount.into()),
            metadata: value.metadata.map(Into::into),
        }
    }
}
