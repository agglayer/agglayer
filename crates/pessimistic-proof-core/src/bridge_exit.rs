#![allow(clippy::too_many_arguments)]

use agglayer_primitives::{address, Address, U256};
use hex_literal::hex;
use serde::{Deserialize, Serialize};

use crate::keccak::{digest::Digest, keccak256_combine};

pub const L1_NETWORK_ID: NetworkId = 0;
pub const L1_ETH: TokenInfo = TokenInfo {
    origin_network: L1_NETWORK_ID,
    origin_token_address: address!("0000000000000000000000000000000000000000"),
};

/// Encapsulates the information to uniquely identify a token on the origin
/// network.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Copy)]
pub struct TokenInfo {
    /// Network which the token originates from
    pub origin_network: NetworkId,
    /// The address of the token on the origin network
    pub origin_token_address: Address,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum LeafType {
    Transfer = 0,
    Message = 1,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
#[error("Invalid leaf type number")]
pub struct LeafTypeFromU8Error;

impl TryFrom<u8> for LeafType {
    type Error = LeafTypeFromU8Error;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Transfer),
            1 => Ok(Self::Message),
            _ => Err(LeafTypeFromU8Error),
        }
    }
}

/// Represents a token bridge exit from the network.
// TODO: Change it to an enum depending on `leaf_type`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BridgeExit {
    pub leaf_type: LeafType,

    /// Unique ID for the token being transferred.
    pub token_info: TokenInfo,

    /// Network which the token is transferred to
    pub dest_network: NetworkId,
    /// Address which will own the received token
    pub dest_address: Address,

    /// Token amount sent
    pub amount: U256,

    pub metadata: Option<Digest>,
}

const EMPTY_METADATA_HASH: Digest = Digest(hex!(
    "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
));

pub fn bridge_exit_hasher(
    leaf_type: u8,
    origin_network: u32,
    origin_token_address: Address,
    dest_network: u32,
    dest_address: Address,
    amount: U256,
    metadata: Option<Digest>,
) -> Digest {
    keccak256_combine([
        [leaf_type].as_slice(),
        &u32::to_be_bytes(origin_network),
        origin_token_address.as_slice(),
        &u32::to_be_bytes(dest_network),
        dest_address.as_slice(),
        &amount.to_be_bytes::<32>(),
        &metadata.unwrap_or(EMPTY_METADATA_HASH).0,
    ])
}

impl BridgeExit {
    /// Hashes the [`BridgeExit`] to be inserted in a
    /// [`crate::local_exit_tree::LocalExitTree`].
    pub fn hash(&self) -> Digest {
        bridge_exit_hasher(
            self.leaf_type as u8,
            self.token_info.origin_network,
            self.token_info.origin_token_address,
            self.dest_network,
            self.dest_address,
            self.amount,
            self.metadata,
        )
    }

    pub fn is_message(&self) -> bool {
        self.leaf_type == LeafType::Message
    }

    /// Returns the [`TokenInfo`] considered for the the given amount.
    /// The amount corresponds to L1 ETH if the bridge exit is a message.
    pub fn amount_token_info(&self) -> TokenInfo {
        match self.leaf_type {
            LeafType::Message => L1_ETH,
            LeafType::Transfer => self.token_info,
        }
    }
}

pub type NetworkId = u32;
