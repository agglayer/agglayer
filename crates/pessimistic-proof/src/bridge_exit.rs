#![allow(clippy::too_many_arguments)]

use std::{fmt::Display, ops::Deref};

use agglayer_primitives::{address, Address, U256};
use hex_literal::hex;
use serde::{Deserialize, Serialize};

use crate::keccak::{digest::Digest, keccak256, keccak256_combine};

pub(crate) const L1_NETWORK_ID: NetworkId = NetworkId(0);
pub(crate) const L1_ETH: TokenInfo = TokenInfo {
    origin_network: L1_NETWORK_ID,
    origin_token_address: address!("0000000000000000000000000000000000000000"),
};

/// Encapsulates the information to uniquely identify a token on the origin
/// network.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Copy)]
pub struct TokenInfo {
    /// Network which the token originates from
    pub origin_network: NetworkId,
    /// The address of the token on the origin network
    pub origin_token_address: Address,
}

impl TokenInfo {
    /// Computes the Keccak digest of [`TokenInfo`].
    pub fn hash(&self) -> Digest {
        keccak256_combine([
            &self.origin_network.to_be_bytes(),
            self.origin_token_address.as_slice(),
        ])
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum LeafType {
    Transfer = 0,
    Message = 1,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
#[error("Invalid leaf type number")]
pub struct LeafTypeFromU8Error;

impl TryFrom<u8> for LeafType {
    type Error = LeafTypeFromU8Error;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Transfer),
            1 => Ok(Self::Message),
            _ => Err(LeafTypeFromU8Error),
        }
    }
}

/// Represents a token bridge exit from the network.
// TODO: Change it to an enum depending on `leaf_type`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BridgeExit {
    pub leaf_type: LeafType,

    /// Unique ID for the token being transferred.
    pub token_info: TokenInfo,

    /// Network which the token is transferred to
    pub dest_network: NetworkId,
    /// Address which will own the received token
    pub dest_address: Address,

    /// Token amount sent
    pub amount: U256,

    pub metadata: Option<Digest>,
}

const EMPTY_METADATA_HASH: Digest = Digest(hex!(
    "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
));

impl BridgeExit {
    /// Creates a new [`BridgeExit`].
    pub fn new(
        leaf_type: LeafType,
        origin_network: NetworkId,
        origin_token_address: Address,
        dest_network: NetworkId,
        dest_address: Address,
        amount: U256,
        metadata: Vec<u8>,
    ) -> Self {
        Self {
            leaf_type,
            token_info: TokenInfo {
                origin_network,
                origin_token_address,
            },
            dest_network,
            dest_address,
            amount,
            metadata: Some(keccak256(metadata.as_slice())),
        }
    }

    /// Hashes the [`BridgeExit`] to be inserted in a
    /// [`crate::local_exit_tree::LocalExitTree`].
    pub fn hash(&self) -> Digest {
        keccak256_combine([
            [self.leaf_type as u8].as_slice(),
            &u32::to_be_bytes(self.token_info.origin_network.into()),
            self.token_info.origin_token_address.as_slice(),
            &u32::to_be_bytes(self.dest_network.into()),
            self.dest_address.as_slice(),
            &self.amount.to_be_bytes::<32>(),
            &self.metadata.unwrap_or(EMPTY_METADATA_HASH).0,
        ])
    }

    pub fn is_transfer(&self) -> bool {
        self.leaf_type == LeafType::Transfer
    }
    pub fn is_message(&self) -> bool {
        self.leaf_type == LeafType::Message
    }

    /// Returns the [`TokenInfo`] considered for the the given amount.
    /// The amount corresponds to L1 ETH if the bridge exit is a message.
    pub fn amount_token_info(&self) -> TokenInfo {
        match self.leaf_type {
            LeafType::Message => L1_ETH,
            LeafType::Transfer => self.token_info,
        }
    }
}

#[derive(
    Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Hash,
)]
pub struct NetworkId(u32);

impl Display for NetworkId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl NetworkId {
    pub const BITS: usize = u32::BITS as usize;
    pub fn new(value: u32) -> Self {
        Self(value)
    }
}

impl From<u32> for NetworkId {
    fn from(value: u32) -> Self {
        Self(value)
    }
}

impl From<NetworkId> for u32 {
    fn from(value: NetworkId) -> Self {
        value.0
    }
}

impl Deref for NetworkId {
    type Target = u32;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::local_exit_tree::{hasher::Keccak256Hasher, LocalExitTree};

    #[test]
    fn test_deposit_hash() {
        let mut deposit = BridgeExit::new(
            LeafType::Transfer,
            0.into(),
            Address::default(),
            1.into(),
            Address::default(),
            U256::default(),
            vec![],
        );

        let amount_bytes = hex::decode("8ac7230489e80000").unwrap_or_default();
        deposit.amount = U256::try_from_be_slice(amount_bytes.as_slice()).unwrap();

        let dest_addr = hex::decode("c949254d682d8c9ad5682521675b8f43b102aec4").unwrap_or_default();
        deposit.dest_address.copy_from_slice(&dest_addr);

        let leaf_hash = deposit.hash();
        assert_eq!(
            "22ed288677b4c2afd83a6d7d55f7df7f4eaaf60f7310210c030fd27adacbc5e0",
            hex::encode(leaf_hash)
        );

        let mut dm = LocalExitTree::<Keccak256Hasher>::new();
        dm.add_leaf(leaf_hash).unwrap();
        let dm_root = dm.get_root();
        assert_eq!(
            "5ba002329b53c11a2f1dfe90b11e031771842056cf2125b43da8103c199dcd7f",
            hex::encode(dm_root)
        );
    }
}
