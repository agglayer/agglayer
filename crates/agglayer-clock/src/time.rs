use std::{
    num::NonZeroU64,
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc,
    },
    time::Duration,
};

use chrono::{DateTime, Utc};
use tokio::{
    sync::broadcast,
    time::{interval_at, Instant},
};
use tokio_util::sync::CancellationToken;
use tracing::{debug, error};

use crate::{Clock, ClockRef, Error, Event, BROADCAST_CHANNEL_SIZE};

/// Time based [`Clock`] implementation.
///
/// Simulate blockchain block production by increasing Block height by 1 every
/// second. Epoch duration can be configured when creating the Clock.
pub struct TimeClock {
    genesis: DateTime<Utc>,
    current_block: Arc<AtomicU64>,
    epoch_duration: NonZeroU64,
    current_epoch: Arc<AtomicU64>,
}

#[async_trait::async_trait]
impl Clock for TimeClock {
    async fn spawn(mut self, cancellation_token: CancellationToken) -> Result<ClockRef, Error> {
        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let clock_ref = ClockRef {
            sender: sender.clone(),
            current_epoch: self.current_epoch.clone(),
            block_height: self.current_block.clone(),
        };

        // Spawn the Clock task directly
        tokio::spawn(async move {
            self.run(sender, cancellation_token).await;
        });

        Ok(clock_ref)
    }
}

impl TimeClock {
    /// Create a new [`TimeClock`] instance based on the current datetime and an
    /// Epoch.
    pub fn new_now(epoch_duration: NonZeroU64) -> Self {
        Self::new(Utc::now(), epoch_duration)
    }

    /// Create a new [`TimeClock`] instance based on a genesis datetime and an
    /// Epoch duration.
    pub fn new(genesis: DateTime<Utc>, epoch_duration: NonZeroU64) -> Self {
        Self {
            genesis,
            current_block: Arc::new(AtomicU64::new(0)),
            epoch_duration,
            current_epoch: Arc::new(AtomicU64::new(0)),
        }
    }

    /// Run the Clock task.
    async fn run(
        &mut self,
        sender: broadcast::Sender<Event>,
        cancellation_token: CancellationToken,
    ) {
        let mut interval = interval_at(Instant::now(), Duration::from_secs(1));

        // Compute the current Block height and Epoch number
        let current_block = self.update_block_height();
        let current_epoch = self.calculate_epoch_number(current_block);

        // Use compare_exchange to ensure the initial current_epoch is 0
        if let Err(stored_value) = self.current_epoch.compare_exchange(
            0,
            current_epoch,
            Ordering::AcqRel,
            Ordering::Relaxed,
        ) {
            let error_message = format!(
                "The current_epoch has already been modified. Shutting down the Clock task. \
                 Stored value: {}",
                stored_value
            );
            #[cfg(not(test))]
            {
                error!("{}", error_message);
                std::process::exit(1);
            }
            #[cfg(test)]
            panic!("{}", error_message);
        }

        loop {
            tokio::select! {
                _ = cancellation_token.cancelled() => {
                    debug!("Clock task cancelled");
                    break;
                }
                _ = interval.tick() => {

                    // Increase the Block height by 1.
                    // The `fetch_add` method returns the previous value, so we need to add 1 to it
                    // to get the current Block height.
                    if let Some(current_block) = self
                        .current_block
                            .fetch_add(1, Ordering::Release)
                            .checked_add(1)
                    {
                        // If the current Block height is a multiple of the Epoch duration,
                        // the current Epoch has ended. In this case, we need to update the
                        // new Epoch number and send an `EpochEnded` event to the subscribers.
                        if current_block % self.epoch_duration == 0 {
                            match self.update_epoch_number() {
                                Ok(epoch_ended) => {
                                    _ = sender.send(Event::EpochEnded(epoch_ended));
                                }
                                Err((current_epoch, expected)) => {
                                    error!(
                                        "Unexpected error computing the current Epoch: current_epoch={}, \
                                        expected_epoch={}, current_block={}",
                                        current_epoch, expected, current_block
                                    );
                                    cancellation_token.cancel();
                                }
                            }
                        }
                    } else {
                       error!("Block height overflowed the u64 limit. \
                           This is an unexpected situation and could lead to unexpected behavior. \
                           Please report this issue to the developers. https://github.com/agglayer/agglayer/issues/new \
                           The node will now kill itself to prevent further damage.");

                        cancellation_token.cancel();
                    }
                }
            }
        }
    }

    /// Updates the Block height of this [`TimeClock`].
    ///
    /// This method is used to update the Block height based on the
    /// genesis datetime and the current datetime.
    ///
    /// The Block height is the number of seconds since the genesis datetime.
    fn update_block_height(&mut self) -> u64 {
        let block_height = self.calculate_block_height();

        self.current_block.store(block_height, Ordering::Release);

        block_height
    }

    /// Updates the current Epoch of this [`TimeClock`].
    ///
    /// This method is used to update the current Epoch number based on the
    /// Block height and the Epoch duration.
    ///
    /// To define the current Epoch number, the Epoch duration divides the Block
    /// height.
    fn update_epoch_number(&mut self) -> Result<u64, (u64, u64)> {
        let current_block = self.current_block.load(Ordering::Acquire);

        let current_epoch = self.calculate_epoch_number(current_block);
        let expected_epoch = current_epoch.saturating_sub(1);

        match self.current_epoch.compare_exchange(
            expected_epoch,
            current_epoch,
            Ordering::Acquire,
            Ordering::Relaxed,
        ) {
            Ok(previous) => Ok(previous),
            Err(stored) => Err((stored, expected_epoch)),
        }
    }

    /// Calculate an Epoch number based on a Block number.
    fn calculate_epoch_number(&self, from_block: u64) -> u64 {
        from_block / self.epoch_duration
    }

    /// Calculate the Block height.
    fn calculate_block_height(&self) -> u64 {
        std::cmp::max(
            Utc::now()
                .naive_utc()
                .signed_duration_since(self.genesis.naive_utc())
                .num_seconds(),
            0,
        ) as u64
    }
}

#[cfg(test)]
mod tests {
    use std::{num::NonZeroU64, sync::atomic::Ordering};

    use chrono::{Duration, Utc};
    use tokio::sync::broadcast;
    use tokio_util::sync::CancellationToken;

    use crate::{Clock, ClockRef, Event, TimeClock, BROADCAST_CHANNEL_SIZE};

    #[tokio::test]
    async fn test_time_clock() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let clock = TimeClock::new(genesis, NonZeroU64::new(5).unwrap());

        let token = CancellationToken::new();
        let clock_ref = clock.spawn(token.clone()).await.unwrap();

        let mut recv = clock_ref.subscribe().unwrap();

        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(6)));
        assert_eq!(clock_ref.current_epoch(), 7);
        assert!(clock_ref.current_block_height() >= 30);
    }

    #[tokio::test]
    async fn test_time_clock_catchup() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let clock = TimeClock::new(genesis, NonZeroU64::new(2).unwrap());

        let token = CancellationToken::new();
        let clock_ref = clock.spawn(token.clone()).await.unwrap();

        let mut recv = clock_ref.subscribe().unwrap();
        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(15)));
        assert!(recv.try_recv().is_err());
        assert_eq!(clock_ref.current_epoch(), 16);
        assert!(clock_ref.current_block_height() >= 30);
        tokio::time::sleep(std::time::Duration::from_secs(5)).await;

        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(16)));
        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(17)));

        assert_eq!(clock_ref.current_epoch(), 18);
        assert!(clock_ref.current_block_height() >= 35);
    }

    #[tokio::test]
    async fn test_time_clock_overflow() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let mut clock = TimeClock::new(genesis, NonZeroU64::new(2).unwrap());
        let blocks = clock.current_block.clone();
        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let clock_ref = ClockRef {
            sender: sender.clone(),
            current_epoch: clock.current_epoch.clone(),
            block_height: clock.current_block.clone(),
        };

        let token = CancellationToken::new();
        let mut fut = Box::pin(clock.run(sender, token.clone()));
        let mut recv = clock_ref.subscribe().unwrap();

        _ = futures::poll!(&mut fut);
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;

        _ = futures::poll!(&mut fut);
        assert_eq!(recv.try_recv(), Ok(Event::EpochEnded(15)));

        // Overflow the block height on next poll
        blocks.store(u64::MAX - 1, std::sync::atomic::Ordering::SeqCst);

        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        _ = futures::poll!(&mut fut);

        assert!(token.is_cancelled());
    }

    #[tokio::test]
    #[should_panic]
    async fn test_initial_epoch_update_error() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let mut clock = TimeClock::new(genesis, NonZeroU64::new(5).unwrap());
        clock.current_epoch.store(1, Ordering::Relaxed);

        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let token = CancellationToken::new();
        let _ = clock.run(sender, token).await;
    }
}
