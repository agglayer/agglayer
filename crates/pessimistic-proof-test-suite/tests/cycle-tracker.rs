use std::time::Duration;

use pessimistic_proof::bridge_exit::BridgeExit;
use pessimistic_proof_test_suite::{forest::Forest, runner::Runner, sample_data as data};
use tracing::{debug, info};

#[rstest::rstest]
#[timeout(Duration::from_secs(30))]
#[case::empty(Forest::new([]), std::iter::empty())]
fn sanity_check(#[case] state: Forest, #[case] bridge_exits: impl Iterator<Item = BridgeExit>) {
    cycles_on_sample_inputs_inner(state, bridge_exits);
}

#[rstest::rstest]
#[case::s01_n000(data::sample_state_01(), std::iter::empty())]
#[case::s01_n001(data::sample_state_01(), data::sample_bridge_exits_01().take(1))]
#[case::s01_n002(data::sample_state_01(), data::sample_bridge_exits_01().take(2))]
#[case::s01_n020(data::sample_state_01(), data::sample_bridge_exits_01().take(20))]
#[case::s01_n100(data::sample_state_01(), data::sample_bridge_exits_01().take(100))]
#[case::s01_full(data::sample_state_01(), data::sample_bridge_exits_01())]
#[ignore = "Too expensive to run by default"]
fn cycles_on_sample_inputs(
    #[case] state: Forest,
    #[case] bridge_exits: impl Iterator<Item = BridgeExit>,
) {
    cycles_on_sample_inputs_inner(state, bridge_exits);
}

fn cycles_on_sample_inputs_inner(
    mut state: Forest,
    bridge_exits: impl Iterator<Item = BridgeExit>,
) {
    sp1_sdk::utils::setup_logger();

    let withdrawals = bridge_exits
        .map(|be| (be.token_info, be.amount))
        .collect::<Vec<_>>();
    let n_exits = withdrawals.len();

    let old_state = state.local_state();
    let (certificate, signer) = state.clone().apply_events(&[], &withdrawals);

    let multi_batch_header = state
        .state_b
        .apply_certificate(
            &certificate,
            signer,
            certificate.l1_info_root().unwrap().unwrap_or_default(),
        )
        .unwrap();

    let (new_roots, stats) = Runner::new()
        .execute(&old_state, &multi_batch_header)
        .expect("execution failed");

    // Double check the roots match what is calculated by the proof-external state.
    state.assert_output_matches(&new_roots);

    debug!("full execution stats:\n{stats}");
    debug!("result: {new_roots:?}");

    let cycles = stats.total_instruction_count();
    let syscalls = stats.total_syscall_count();
    info!("execution summary: n={n_exits}, cycles={cycles}, syscalls={syscalls}");
}
