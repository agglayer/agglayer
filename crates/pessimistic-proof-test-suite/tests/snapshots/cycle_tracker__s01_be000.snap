---
source: crates/pessimistic-proof-test-suite/tests/cycle-tracker.rs
expression: stats
---
opcode counts (638992 total instructions):
    159708 add
    103759 sw
     93608 lw
     49952 lbu
     35234 sb
     28659 xor
     23055 bne
     19126 beq
     16840 bltu
     14653 jalr
     14544 sltu
     13040 and
     11401 sub
     10303 or
      9869 srl
      9796 sll
      7991 auipc
      4041 jal
      3771 sra
      2843 lb
      2474 mul
      2097 mulhu
      1063 ecall
       841 bgeu
       308 sh
         7 lh
         4 lhu
         4 blt
         1 divu
syscall counts (1063 total syscall instructions):
    512 secp256k1_double
    263 keccak_permute
    259 secp256k1_add
      8 commit
      8 commit_deferred_proofs
      5 hint_len
      5 hint_read
      2 write
      1 halt
